src/cmd/ksh93/sh/subshell.c

```diff
@@ -252,7 +252,7 @@ Namval_t *sh_assignok(register Namval_t *np,int add)
 	if(sp->subshare)
 		return(np);
 	/* don't bother with this */
-	if(!sp->shpwd || np==SH_LEVELNOD || np==L_ARGNOD || np==SH_SUBSCRNOD || np==SH_NAMENOD)
+	if(!sp->shpwd || (add != 2 && (np==SH_LEVELNOD || np==L_ARGNOD || np==SH_SUBSCRNOD || np==SH_NAMENOD)))
```

Comment:
I would like to extend the comment before the `sh_assignok()` function definition to better explain what it does and what the `add` parameter means (including the new value).

I'm still not sure I get it completely. Does the following look correct to you?

```c
/*
 * This routine will make a copy of the given node in the
 * layer created by the most recent virtual subshell if the
 * node hasn't already been copied.
 *
 * add == 0:    Move the node pointer from the parent shell to the current virtual subshell.
 * add == 1:    Create a copy of the node pointer in the current virtual subshell.
 * add == 2:    For restoring variable state when exiting virtual subshell. Same as 1, but also
 *              disable optimization for ${.sh.level}, $_, ${.sh.subscript} and ${.sh.name}.
 */
```

Reply:
The comment is a good idea (added to the pull request), although I have revised it with a different explanation.